modules/libcom/src/iocsh/libComRegister.c

```diff
@@ -28,6 +29,13 @@
 #include "freeList.h"
 #include "libComRegister.h"
 
+struct dirStackEntry {
+    ELLNODE node;
+	char path[];
+};
+
+static ELLLIST dirStack = ELLLIST_INIT;
```

Comment:
IOC shell functions may run concurrently.  For basic user sanity, I don't think it is reasonably to have all threads share a common stack.

I think it would be better to add this stack to the per-thread `struct iocshContext`.

Reply:
Also, since `iocsh.cpp` is C++, maybe `std::list<std::string>` would avoid some effort, and avoid memory leaks?

Reply:
Do individual Posix, Windows and RTEMS-score threads each have their own record of the CWD, or do they share one CWD for the whole process? I'm pretty sure VxWorks only has one CWD, although I'm be okay ignoring that target if it's unique. If it isn't and concurrent iocsh instances are calling `cd`/`pushd`/`popd` there are going to be some major problems with relative paths. We tell people never to use relative paths for the access security configuration file because someone running a `cd` command would cause a DoS if/when that file gets reloaded (iocStats loads a PV for doing that).

Reply:
@anjohnson I am not referring to `getcwd()`, but only to this iocsh directory stack.  I acknowledge that it is not possible to avoid races with concurrent `chdir()`.  I understand though that one of the goals of this change is to bring a measure of sanity.  I think that a per-thread stack will help with this.

Reply:
I've added this to `iocshContext` and switched to a `std::list` of `std::string`s